### PR TITLE
Update snapshot storing, replace sled db with the already used diesel sql database

### PR DESCRIPTION
**Pull request by Mr-Leshiy**
It is a preparation before actual `/api/v0/user/{id}:`endpoint implementation.
- remove sled dependency, replace its usage with diesel
- store all information from the snapshot and putting it into the relational database

### CODE REVIEW

**Comment by Mr-Leshiy** [comment]
firstly need to merge https://github.com/input-output-hk/catalyst-toolbox/pull/167

File: vit-servicing-station-lib/migrations/2020-05-22-112032_setup_db/up.sql [hunk 1]
```diff
@@ -145,6 +145,31 @@ create table votes (
...
+create table voters (
+    voting_key TEXT NOT NULL,
+    voting_power BIGINT NOT NULL,
+    voting_group TEXT NOT NULL,
+    snapshot_tag TEXT NOT NULL,
+    PRIMARY KEY(voting_key, voting_group, snapshot_tag)
+    FOREIGN KEY(snapshot_tag) REFERENCES snapshots(tag) ON DELETE CASCADE
+);
+
+create table contributors (
```
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply vit-servicing-station-lib/migrations/2020-05-22-112032_setup_db/up.sql:162]: agree, actually in my implementation it is totally duplicates voting_group from the `voter` table

**Comment by Mr-Leshiy** [comment]
@dkijania Initially I have the following ideas

**Voter** it is who actually votes in the jormungandr blockchain and corresponds `hir` field https://github.com//blob/0d43013f6eae212fc97eccab60cb97cda2d3d3fb/snapshot-lib/src/lib.rs#L51-L52. But also we should distinguish them by different snapshots, because in two different snapshots which corresponds to different funds we can have the same voter and probably with the same voting_power and voter_group fields. That is why I have added `snapshot_tag` column as well.

**Contributor** it is who delegates his voting power to different voters (representatives). So each **Contributor** could has different voters with the different voting groups, note that I am assuming that contributor could delegates his voting power to certain voter with the certain voter group (but voter can be applied to another voter groups which would be different with that contributor want to delegates). That is why I have added `voting_key` and `voting_group` columns here. `snapshot_tag` used analogically as **Voter** used

File: vit-servicing-station-lib/migrations/2020-05-22-112032_setup_db/up.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (cong-or) [reply vit-servicing-station-lib/migrations/2020-05-22-112032_setup_db/up.sql:162]: @Mr-Leshiy not sure about this either. However, Im sure we can change it down the line anyway..

File: vit-servicing-station-lib/src/v0/endpoints/snapshot/routes.rs [hunk 2]
```diff
@@ -25,10 +24,9 @@ pub fn filter(
 }
 
 pub fn update_filter(
-    context: UpdateHandle,
+    context: SharedContext,
```
Comment (cong-or) [inline vit-servicing-station-lib/src/v0/endpoints/snapshot/routes.rs:27]: why did you change this? not sure what the purpose of this crate `v0::context::SharedContext` is?

File: vit-servicing-station-lib/src/db/models/snapshot.rs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    /// Tag - a unique identifier of the current snapshot
+    pub tag: String,
+    /// Timestamp for the latest update of the current snapshot
+    #[serde(alias = "lastUpdated")]
+    #[serde(deserialize_with = "crate::utils::serde::deserialize_unix_timestamp_from_rfc3339")]
+    #[serde(serialize_with = "crate::utils::serde::serialize_unix_timestamp_as_rfc3339")]
+    pub last_updated: i64,
+}
+
+impl Queryable<snapshots::SqlType, Db> for Snapshot {
```
Comment (cameron1024) [inline vit-servicing-station-lib/src/db/models/snapshot.rs:20]: Is there a reason we're not deriving this?

File: vit-servicing-station-lib/src/db/models/snapshot.rs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+    fn build(row: Self::Row) -> Self {
+        Self {
+            tag: row.0,
+            last_updated: row.1,
+        }
+    }
+}
+
+impl Insertable<snapshots::table> for Snapshot {
```
Comment (cameron1024) [inline vit-servicing-station-lib/src/db/models/snapshot.rs:36]: Same here, does deriving not work?

File: vit-servicing-station-lib/src/db/models/snapshot.rs [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        (
+            snapshots::tag.eq(self.tag),
+            snapshots::last_updated.eq(self.last_updated),
+        )
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Voter {
+    #[serde(alias = "votingKey")]
```
Comment (cameron1024) [inline vit-servicing-station-lib/src/db/models/snapshot.rs:52]: We could put `#[serde(rename_all = "camelCase")]` on the struct and remove these aliases?

File: vit-servicing-station-lib/src/db/models/snapshot.rs [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+            voters::voting_power.eq(self.voting_power),
+            voters::voting_group.eq(self.voting_group),
+            voters::snapshot_tag.eq(self.snapshot_tag),
+        )
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Contributor {
+    #[serde(alias = "stakePublicKey")]
```
Comment (cameron1024) [inline vit-servicing-station-lib/src/db/models/snapshot.rs:104]: Same again with `renameAll`

File: vit-servicing-station-lib/src/db/models/snapshot.rs [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    #[serde(alias = "votingKey")]
+    pub voting_key: String,
+    #[serde(alias = "votingGroup")]
+    pub voting_group: String,
+    #[serde(alias = "snapshotTag")]
+    pub snapshot_tag: String,
+}
+
+impl Queryable<contributors::SqlType, Db> for Contributor {
+    type Row = (
```
Comment (cameron1024) [inline vit-servicing-station-lib/src/db/models/snapshot.rs:118]: Derive this too?

File: vit-servicing-station-lib/src/db/models/snapshot.rs [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+            reward_address: row.1,
+            value: row.2,
+            voting_key: row.3,
+            voting_group: row.4,
+            snapshot_tag: row.5,
+        }
+    }
+}
+
+impl Insertable<contributors::table> for Contributor {
```
Comment (cameron1024) [inline vit-servicing-station-lib/src/db/models/snapshot.rs:145]: Derive this too?

File: vit-servicing-station-lib/src/v0/endpoints/snapshot/handlers.rs [hunk 9]
```diff
@@ -47,23 +42,18 @@ pub async fn get_voters_info(
...
     }
 }
 
 #[tracing::instrument(skip(context))]
 pub async fn get_tags(context: SharedContext) -> Result<impl Reply, Rejection> {
-    match context.get_tags().map(|tags| warp::reply::json(&tags)) {
+    match super::get_tags(context)
+        .await
+        .map(|tags| warp::reply::json(&tags))
+    {
```
Comment (cameron1024) [inline vit-servicing-station-lib/src/v0/endpoints/snapshot/handlers.rs:54]: Slight style nit: can we pull this expression into a variable, and the have `match tags {` ?

**Review by cong-or — APPROVED** [review]
LGTM

Not 100% sure about voters table. However, it doesn't seems like a big issue if you need to change it.

File: vit-servicing-station-lib/src/db/queries/snapshot.rs [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    db::{
+        models::snapshot::{Contributor, Snapshot, Voter},
+        schema::{contributors, snapshots, voters},
+        DbConnection, DbConnectionPool,
+    },
+    v0::errors::HandleError,
+};
+use diesel::{ExpressionMethods, Insertable, QueryDsl, RunQueryDsl};
+
+pub async fn query_all_snapshots(pool: &DbConnectionPool) -> Result<Vec<Snapshot>, HandleError> {
```
Comment (cong-or) [inline vit-servicing-station-lib/src/db/queries/snapshot.rs:11]: how often is this called? what is the context for its use?

File: vit-servicing-station-lib/src/db/models/snapshot.rs (hunk 4 above)
Comment (Mr-Leshiy) [reply vit-servicing-station-lib/src/db/models/snapshot.rs:16]: Unfortunately does not work

File: vit-servicing-station-lib/src/db/models/snapshot.rs (hunk 5 above)
Comment (Mr-Leshiy) [reply vit-servicing-station-lib/src/db/models/snapshot.rs]: fixed

File: vit-servicing-station-lib/src/db/models/snapshot.rs (hunk 3 above)
Comment (Mr-Leshiy) [reply vit-servicing-station-lib/src/db/models/snapshot.rs]: Good catch ! Updated it !

File: vit-servicing-station-lib/src/v0/endpoints/snapshot/routes.rs (hunk 2 above)
Comment (Mr-Leshiy) [reply vit-servicing-station-lib/src/v0/endpoints/snapshot/routes.rs:27]: I have replaced and removed `sled` database usage with the diesel. `UpdateHandle` stored the context of the `sled` database, `v0::context::SharedContext` stores diesel database context, so that is why I have replaced it

File: vit-servicing-station-lib/src/db/queries/snapshot.rs (hunk 10 above)
Comment (Mr-Leshiy) [reply vit-servicing-station-lib/src/db/queries/snapshot.rs:11]: It is used in this places https://github.com/input-output-hk/vit-servicing-station/blob/46e15f4f3f3e40aa19720ae7a322781e53bb370d/vit-servicing-station-lib/src/v0/endpoints/snapshot/mod.rs#L85
